ls1intum/Artemis
# `General`: Increase interactivity of active students line charts and course exercises average bar chart

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


This PR increases the interactivity for several charts:

- The line chart on the course detail page
- The line chart in Course administration
- The bar chart displaying the average student score of each exercise in course statistics

### Description

It is now possible to perform some more interaction with these chats:

- Next to the title of the line charts, a Tooltip is now placed that explains the term "active students"
- The title of the line charts are now clickable
- The line chart on the course detail page is now displayable in full screen mode
- By clicking on a bar, the user is delegated to the corresponding exercise statistics page

Besides that, some minor adjustments in the diagrams are included in this PR as well:

- The lines in the line chart are now smoothed in order to present a trend-line.
- The size difference within the tooltips is now a bit reduced
- The line chart on the course detail page now displays one additional week in the view
- The tooltips for specific points in the line chart now contain the relative value represented by the point as number
- The line chart on the course detail page contains the average over the displayed time frame as reference line that is toggleable
- For the bar chart on the course exercise, now a first color coding is introduced: If the average score is higher than 90% the bar is displayed in green, it it is below 25% it is displayed in red
- The tooltips of the bar chart now contain the exercise type of the corresponding exercise

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course with some exercises and participation

1. Log in to Artemis
2. Navigate to Course Administration
3. Inspect the line chart there. Have a look at the tooltips in the chart and the explanation tooltip next to the title. Also check the translation of the explanation tooltip.
4. Click on the chart title. You should be delegated to the corresponding course detail page.
5. Inspect the more detailed line chart there. Again, check the tooltips and the explanation translation. Make sure that the reference line disappears if the check box is unchecked.
6. Also inspect the appearance of the chart in full screen mode.

-  **Important**: There is a remaining problem with the fullscreen mode of the chart: The tooltips are not visible in full screen. This is due to the underlying mechanics of `ngx-charts` (at least as far as I can tell). I found [a PR](https://github.com/swimlane/ngx-charts/pull/1114) in the GitHub project of the framework that seems to tackle this issue. But as this PR is from Freburary of 2019 at it is still open, I do not have much hope that this is still maintained. 

7. Click on the chart title again. You should be delegated to the course statistics.
8. Check the first bar chart there. Make sure that the color coding works as expected. Also check the tooltips if the exercise type is matching and the translations are appropriate/ aligned with translations at oder places in Artemis.
9. Click on a bar of your choice. You should be delegated to the corresponding exercise statistics page.

                       

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage





statistics-average-score-graph.component.ts: 72.72% | 93.87%
course-detail-line-chart.component.ts: 68.42% |  92.85% 
### Screenshots



The line chart on course administration:
![image](https://user-images.githubusercontent.com/80622272/147854975-8510d3b1-d150-4579-bab6-b00d460d4f4d.png)

The line chart on course detail page:
![image](https://user-images.githubusercontent.com/80622272/147884260-3b57041f-639d-40bd-ac65-b1414eddc321.png)

The bar chart in the course statistics:
![image](https://user-images.githubusercontent.com/80622272/147855016-1c61bd3e-fde7-4aba-9052-6a4755544bd2.png)
